go/vt/vtgr/db/mysql.go

```diff
@@ -0,0 +1,535 @@
+	"vitess.io/vitess/go/vt/orchestrator/external/golib/sqlutils"
+	"vitess.io/vitess/go/vt/orchestrator/inst"
+)
+
+var (
+	configFile     = flag.String("db_config", "", "db config file name")
+	dbFlavor       = flag.String("db_flavor", "MySQL56", "mysql flavor override")
+	mysqlGroupPort = flag.Int("gr_port", 33061, "port to bootstrap a mysql group")
+
+	// ErrGroupSplitBrain is the error when mysql group is split-brain
+	ErrGroupSplitBrain = errors.New("group has split brain")
+	// ErrGroupBackoffError is either the transient error or network partition from the group
+	ErrGroupBackoffError = errors.New("group backoff error")
+	// ErrGroupOngoingBootstrap is the error when a bootstrap is in progress
+	ErrGroupOngoingBootstrap = errors.New("group ongoing bootstrap")
+	// ErrGroupInactive is the error when mysql group is inactive unexpectedly
+	ErrGroupInactive = errors.New("group is inactive")
+	// ErrInvalidInstance is the error when the instance key has empty hostname
+	ErrInvalidInstance = errors.New("invalid mysql instance key")
+)
+
+// Agent is used by vtgr to interact with Mysql
+type Agent interface {
+	// BootstrapGroupLocked bootstraps a mysql group
+	// the caller should grab a lock before
+	BootstrapGroupLocked(instanceKey *inst.InstanceKey) error
+
+	// StopGroupLocked stops a mysql group
+	StopGroupLocked(instanceKey *inst.InstanceKey) error
+
+	// JoinGroupLocked puts an instance into a mysql group based on primary instance
+	// the caller should grab a lock before
+	JoinGroupLocked(instanceKey *inst.InstanceKey, primaryKey *inst.InstanceKey) error
+
+	// SetReadOnly set super_read_only variable
+	// https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_super_read_only
+	SetReadOnly(instanceKey *inst.InstanceKey, readOnly bool) error
+
+	// FetchApplierGTIDSet fetches the GTID set from group_replication_applier channel
+	FetchApplierGTIDSet(instanceKey *inst.InstanceKey) (mysql.GTIDSet, error)
+
+	// Failover move the mysql primary to the node defined by memberUUID
+	Failover(instance *inst.InstanceKey) error
+
+	// FetchGroupView fetches group related information
+	FetchGroupView(alias string, instanceKey *inst.InstanceKey) (*GroupView, error)
+}
+
+// MemberState is member state
+type MemberState int
+
+// MemberRole is member role
+type MemberRole int
+
+const (
+	UNKNOWNSTATE MemberState = iota
+	OFFLINE
+	UNREACHABLE
+	RECOVERING
+	ONLINE
+	ERROR
+)
+
+const (
+	UNKNOWNROLE MemberRole = iota
+	SECONDARY
+	PRIMARY
+)
+
+// GroupMember represents a ROW we get from performance_schema
+type GroupMember struct {
+	HostName string
+	Port     int
+	Role     MemberRole
+	State    MemberState
+	ReadOnly bool
+}
+
+// GroupView is an instance's view for the group
+type GroupView struct {
+	TabletAlias       string
+	MySQLHost         string
+	MySQLPort         int
+	GroupName         string
+	UnresolvedMembers []*GroupMember
+}
+
+// SQLAgentImp implements sqlAgent
+type SQLAgentImp struct {
```

Comment:
naming suggestion: `SQLAgentImpl` or `SqlAgentImpl`.
Also looks like the interface has been renamed to `Agent` but comment still refers to `sqlAgent`.

Reply:
Updated